File: src/canonical.mjs
import { readFileSync } from "node:fs";
import { join } from "node:path";

const TAXONOMY_PATH = join(process.cwd(), "data", "category-taxonomy.json");
const MERCHANTS_PATH = join(process.cwd(), "data", "merchant-aliases.json");

let taxonomyCache;
let merchantCache;

export function normalizeText(input) {
  return String(input || "")
    .normalize("NFKD")
    .replace(/[’‘`´]/g, "'")
    .trim()
    .toLowerCase();
}

export function compactText(input) {
  return normalizeText(input).replace(/[^a-z0-9]+/g, "");
}

export function canonicalizeCategory(value) {
  const taxonomy = loadTaxonomy();
  const normalized = normalizeText(value);
  const compact = compactText(value);
  if (!normalized) return null;

  const legacy = taxonomy.legacyCategoryMap[normalized];
  if (legacy) return legacy;

  for (const category of taxonomy.categories) {
    if (category.id === normalized) return category.id;
    for (const alias of category.aliases || []) {
      const aliasCompact = compactText(alias);
      if (normalizeText(alias) === normalized || (compact && aliasCompact && aliasCompact === compact)) {
        return category.id;
      }
    }
  }

  return null;
}

export function canonicalizeMerchant(value) {
  const normalized = normalizeText(value);
  const compact = compactText(value);
  if (!normalized) return { canonicalMerchant: null, canonicalCategory: null, confidence: 0 };

  for (const merchant of loadMerchants().merchants) {
    for (const alias of merchant.aliases || []) {
      const aliasText = normalizeText(alias);
      const aliasCompact = compactText(alias);
      if (normalized === aliasText || (compact && aliasCompact && compact === aliasCompact)) {
        return {
          canonicalMerchant: merchant.canonical,
          canonicalCategory: merchant.category,
          confidence: 1,
        };
      }
      if (aliasCompact.length >= 4 && compact.includes(aliasCompact)) {
        return {
          canonicalMerchant: merchant.canonical,
          canonicalCategory: merchant.category,
          confidence: 0.9,
        };
      }
    }
  }

  const category = canonicalizeCategory(value);
  if (category) return { canonicalMerchant: null, canonicalCategory: category, confidence: 0 };

  return { canonicalMerchant: value ? String(value).trim() : null, canonicalCategory: null, confidence: 0.4 };
}

export function canonicalizeOffer(offer) {
  const merchant = canonicalizeMerchant(offer.merchant);
  const category =
    merchant.canonicalCategory ||
    canonicalizeCategory(offer.category) ||
    inferCanonicalCategory(offer.merchant, offer.rewardText || offer.reward || offer.sourceText);

  return {
    canonicalMerchant: merchant.canonicalMerchant || String(offer.merchant || "").trim(),
    canonicalCategory: category || "general_shopping",
    categoryConfidence: merchant.canonicalCategory ? merchant.confidence : category ? 0.75 : 0.3,
  };
}

export function classifyQuery(query) {
  const category = canonicalizeCategory(query);
  const merchant = canonicalizeMerchant(query);
  const inferredCategory = category || inferCanonicalCategory(query, "");

  return {
    canonicalMerchant: merchant.confidence >= 0.9 ? merchant.canonicalMerchant : null,
    canonicalCategory: merchant.confidence >= 0.9 ? merchant.canonicalCategory : inferredCategory,
    merchantConfidence: merchant.confidence,
  };
}

export function isKnownCanonicalCategory(value) {
  return Boolean(canonicalizeCategory(value));
}

export function listCanonicalCategoryIds() {
  return loadTaxonomy().categories.map((category) => category.id);
}

function inferCanonicalCategory(merchant = "", rewardText = "") {
  const text = normalizeText(`${merchant} ${rewardText}`);

  if (/\b(restaurant|restaurants|dining|takeout|delivery|coffee|pizza|burger|doordash|resy)\b/.test(text) || /餐厅|吃饭|饭店|咖啡/.test(text)) {
    return "dining";
  }
  if (/\b(hotel|hotels|resort|lodging)\b/.test(text) || /酒店|住宿/.test(text)) return "hotel";
  if (/\b(flight|airline|airfare|airport)\b/.test(text) || /机票|航班/.test(text)) return "airfare";
  if (/\b(travel|cruise|parking|train|airbnb|las vegas|fontainebleau)\b/.test(text) || /旅行|旅游/.test(text)) return "travel";
  if (/\b(grocery|groceries|supermarket|whole foods)\b/.test(text) || /超市|买菜/.test(text)) return "grocery";
  if (/\b(gas|fuel|charging)\b/.test(text) || /加油/.test(text)) return "gas";
  if (/\b(drugstore|pharmacy|cvs|walgreens)\b/.test(text) || /药店/.test(text)) return "drugstore";
  if (/\b(streaming|subscription|netflix|spotify|hulu)\b/.test(text)) return "streaming";
  if (/\b(gym|fitness|pilates)\b/.test(text) || /健身/.test(text)) return "fitness";
  if (/\b(insurance|loan|creditsecure|financial)\b/.test(text) || /保险/.test(text)) return "financial";
  if (/\b(macy's|macys|saks|nordstrom|department store)\b/.test(text) || /百货|商场/.test(text)) return "department_store";
  if (/\b(clothing|clothes|apparel|shoes|sneakers|suit|dress)\b/.test(text) || /衣服|服装|鞋|西装/.test(text)) return "clothing";
  if (/\b(electronics|computer|laptop|phone|appliance)\b/.test(text) || /电器|电脑|手机/.test(text)) return "electronics";
  if (/\b(furniture|hardware|home improvement)\b/.test(text) || /家具|家居|装修/.test(text)) return "home_improvement";
  return null;
}

function loadTaxonomy() {
  if (!taxonomyCache) taxonomyCache = JSON.parse(readFileSync(TAXONOMY_PATH, "utf8"));
  return taxonomyCache;
}

function loadMerchants() {
  if (!merchantCache) merchantCache = JSON.parse(readFileSync(MERCHANTS_PATH, "utf8"));
  return merchantCache;
}

File: src/ccm_py/canonical.py
from __future__ import annotations

import json
import re
from functools import lru_cache
from typing import Any

from .util import ROOT, compact_text, normalize_text


@lru_cache(maxsize=1)
def load_taxonomy() -> dict[str, Any]:
    return json.loads((ROOT / "data" / "category-taxonomy.json").read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def load_merchants() -> dict[str, Any]:
    return json.loads((ROOT / "data" / "merchant-aliases.json").read_text(encoding="utf-8"))


def canonicalize_category(value: Any) -> str | None:
    normalized = normalize_text(value)
    compact = compact_text(value)
    if not normalized:
        return None

    taxonomy = load_taxonomy()
    legacy = taxonomy.get("legacyCategoryMap", {}).get(normalized)
    if legacy:
        return legacy

    for category in taxonomy.get("categories", []):
        if category.get("id") == normalized:
            return category["id"]
        for alias in category.get("aliases", []):
            alias_compact = compact_text(alias)
            if normalize_text(alias) == normalized or (compact and alias_compact and compact == alias_compact):
                return category["id"]
    return None


def canonicalize_merchant(value: Any) -> dict[str, Any]:
    normalized = normalize_text(value)
    compact = compact_text(value)
    if not normalized:
        return {"canonicalMerchant": None, "canonicalCategory": None, "confidence": 0}

    for merchant in load_merchants().get("merchants", []):
        for alias in merchant.get("aliases", []):
            alias_text = normalize_text(alias)
            alias_compact = compact_text(alias)
            if normalized == alias_text or (compact and alias_compact and compact == alias_compact):
                return {
                    "canonicalMerchant": merchant.get("canonical"),
                    "canonicalCategory": merchant.get("category"),
                    "confidence": 1,
                }
            if len(alias_compact) >= 4 and alias_compact in compact:
                return {
                    "canonicalMerchant": merchant.get("canonical"),
                    "canonicalCategory": merchant.get("category"),
                    "confidence": 0.9,
                }

    category = canonicalize_category(value)
    if category:
        return {"canonicalMerchant": None, "canonicalCategory": category, "confidence": 0}
    return {"canonicalMerchant": str(value or "").strip() or None, "canonicalCategory": None, "confidence": 0.4}


def classify_query(query: str) -> dict[str, Any]:
    category = canonicalize_category(query)
    merchant = canonicalize_merchant(query)
    inferred = category or infer_canonical_category(query, "")
    return {
        "canonicalMerchant": merchant["canonicalMerchant"] if merchant["confidence"] >= 0.9 else None,
        "canonicalCategory": merchant["canonicalCategory"] if merchant["confidence"] >= 0.9 else inferred,
        "merchantConfidence": merchant["confidence"],
    }


def canonicalize_offer(offer: dict[str, Any]) -> dict[str, Any]:
    merchant = canonicalize_merchant(offer.get("merchant"))
    category = (
        merchant.get("canonicalCategory")
        or canonicalize_category(offer.get("category"))
        or infer_canonical_category(offer.get("merchant"), offer.get("rewardText") or offer.get("reward_text") or offer.get("sourceText"))
    )
    return {
        "canonicalMerchant": merchant.get("canonicalMerchant") or str(offer.get("merchant") or "").strip(),
        "canonicalCategory": category or "general_shopping",
        "categoryConfidence": merchant.get("confidence") if merchant.get("canonicalCategory") else (0.75 if category else 0.3),
    }


def infer_canonical_category(merchant: Any = "", reward_text: Any = "") -> str | None:
    text = normalize_text(f"{merchant or ''} {reward_text or ''}")
    checks = [
        ("dining", r"\b(restaurant|restaurants|dining|takeout|delivery|coffee|pizza|burger|doordash|resy)\b|餐厅|吃饭|饭店|咖啡"),
        ("hotel", r"\b(hotel|hotels|resort|lodging)\b|酒店|住宿"),
        ("airfare", r"\b(flight|airline|airfare|airport)\b|机票|航班"),
        ("travel", r"\b(travel|cruise|parking|train|airbnb|las vegas|fontainebleau)\b|旅行|旅游"),
        ("grocery", r"\b(grocery|groceries|supermarket|whole foods)\b|超市|买菜"),
        ("gas", r"\b(gas|fuel|charging)\b|加油"),
        ("drugstore", r"\b(drugstore|pharmacy|cvs|walgreens)\b|药店"),
        ("streaming", r"\b(streaming|subscription|netflix|spotify|hulu)\b"),
        ("fitness", r"\b(gym|fitness|pilates)\b|健身"),
        ("financial", r"\b(insurance|loan|creditsecure|financial)\b|保险"),
        ("department_store", r"\b(macy's|macys|saks|nordstrom|department store)\b|百货|商场"),
        ("clothing", r"\b(clothing|clothes|apparel|shoes|sneakers|suit|dress)\b|衣服|服装|鞋|西装"),
        ("electronics", r"\b(electronics|computer|laptop|phone|appliance)\b|电器|电脑|手机"),
        ("home_improvement", r"\b(furniture|hardware|home improvement)\b|家具|家居|装修"),
    ]
    for category, pattern in checks:
        if re.search(pattern, text):
            return category
    return None
